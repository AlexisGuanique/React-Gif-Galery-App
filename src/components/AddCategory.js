import React, { useState } from 'react'
import PropTypes from 'prop-types';


// Creamos componente para aÃ±adir una categoria
export const AddCategory = ( { setCategories } ) => {
    
    // Definimos el estado inicial de nuestro value
    const [InputValue, setInputValue] = useState('');


    // Extraemos el valor a guardar
    const handlInputChange = (e) => {
        // console.log(e.target.value);
        setInputValue(e.target.value);

        // console.log('HandleInputChange llamado')
    }


    // De esta manera evitamos que refresque el formulario de la manera que viene por defecto
    const handleSubmit = (e) => {
        e.preventDefault()

        // console.log('handleSubmit')

        // Validamos que el valor ingresado sea una palabra mayor de 2 letras
        if ( InputValue.trim().length > 1) {

            setCategories( cats => [ InputValue, ...cats ]);
            setInputValue('');
        }
    };
  

    // Lo que renderiza nuestro componente
    return (
        <form onSubmit={ handleSubmit }>
            <p>{ InputValue }</p>
            <input 
                type="text"
                value={ InputValue }
                onChange= { handlInputChange }         
            />
        </form>
  )
};

// De esta manera decimos que la Prop setCategories es requerida para utilizar nuestro componente AddCategory
AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired
};
