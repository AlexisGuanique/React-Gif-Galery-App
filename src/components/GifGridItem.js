import React from 'react';
import { PropTypes } from 'prop-types';

export const GifGridItem = ( {title, url } ) => {

    // console.log( id, title, url );

    return (
        <div className='card animate__animated animate__fadeIn'>
            <img src={ url } alt={ title }/>
            <p className='title'>{ title }</p>
        </div>
    )
};


// Configuracion de los PropTypes, de esta manera indicamos los lineamientos para trabajar con nuestro componente
GifGridItem.protoType = {
    title: PropTypes.string.isRequired,
    url: PropTypes.string.isRequired
};
