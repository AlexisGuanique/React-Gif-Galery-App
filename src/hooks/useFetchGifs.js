import { useEffect, useState } from "react";
import { getGifs } from "../helpers/getGifs";

// CUSTOM HOOK
export const useFetchGifs = ( category ) => {
    // De esta manera manejamos el estado del componente
    const [state, setState] = useState({
        data: [],
        loading: true
    });

    // De esta manera indicamos el efecto para que solo se dispare cuando el componente sea modificado
    useEffect( ()=>{
        
        getGifs( category )
            .then( imgs => {
                    
                setState({
                    data: imgs,
                    loading: false
                });

            }, 2000);
                
        

    },[ category ]);


    

    return state; // { data: [], loading: true }
}