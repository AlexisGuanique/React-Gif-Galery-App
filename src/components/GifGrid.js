import React from 'react'
import { useFetchGifs } from '../hooks/useFetchGifs';
import { GifGridItem } from './GifGridItem';


// Componente que consume la api_key y muestra los gif
export const GifGrid = ( { category } ) => {


    // Utilizamos nuestro custom hook
    const { data:images, loading } = useFetchGifs( category );

    console.log(images);

    return (
        <>
            <h2 className='animate__animated animate__fadeIn'> { category } </h2>

            { loading && <p className='animate__animated animate__flash'>Cargando...</p> }
            
            <div className='card-grid'>
                {
                    images.map( img => (
                        <GifGridItem
                            key={ img.id } 
                            { ...img }
                        />
                    ))
                }
            </div>
        </>
    )
};

