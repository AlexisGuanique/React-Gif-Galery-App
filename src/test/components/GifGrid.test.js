import React from "react";
import '@testing-library/jest-dom';

import { GifGrid } from "../../components/GifGrid";
import { shallow } from 'enzyme';
import { useFetchGifs } from "../../hooks/useFetchGifs";

// Con jest.mock vamos a simular la llamada a cualquier archivo y finjir que tensmoa la informacion necesaria
jest.mock('../../hooks/useFetchGifs')



describe('Pruebas en componente <GifGrid />', () =>{

    // Falsa implementacion del useFetchGifs
    useFetchGifs.mockReturnValue({
        data: [],
        loading: true
    });

    // Variables globales
    const category = 'Dragon Ball Z'
    const wrapper = shallow( <GifGrid category={ category } />)
   


    test('Debe mostrar el componente <GifGrid /> correctamente', () => {
        
        // console.log(wrapper.html() )
        expect( wrapper ).toMatchSnapshot();

    });
    
    test('Debe de mostrar items cuando se cargan imagenes useFetchGifs', () => {

        // Dibujemos el gifs que deberiamos tener
        const gifs = [{
            id: '123',
            url: 'https://localhost/tumama/enpantaletas.jp',
            title: 'La Vieja PEYA'
        },{
            id: '123',
            url: 'https://localhost/tumama/enpantaletas.jp',
            title: 'La Vieja PEYA'
        }];

        ////////////////////////////////////////////////////////////////////////////////////////
        // Vamos a simular que cuando nuestro componente llame al useFetchGifs tenga cierta data

        // obviamente utilizando la variable gifs que dibujamos arriba, ahre
        useFetchGifs.mockReturnValue({
            data: gifs,
            loading: false
        });

        // Llamamos de nuevo al componente para dibujarlo, porque la imagen globan del mismo es de su estado inicial
        const wrapper = shallow( <GifGrid category={ category } />)  
        expect( wrapper ).toMatchSnapshot();

        // Ahora vamos a evaluar que el parrafo donde dice loading no exista, porque obviamente tenemos una imagen 
        expect( wrapper.find('p').exists() ).toBe( false );


        // Ahora vamos a evaluar que exista la llamada del gif a travez del componente GifGridItem
        // Cuantos gifs sean necesarios, por eso gifs.length
        expect( wrapper.find( 'GifGridItem' ).length ).toBe( gifs.length );


    })
})