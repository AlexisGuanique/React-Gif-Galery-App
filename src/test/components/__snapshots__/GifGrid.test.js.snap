// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en componente <GifGrid /> Debe de mostrar items cuando se cargan imagenes useFetchGifs 1`] = `
<Fragment>
  <h2
    className="animate__animated animate__fadeIn"
  >
     
    Dragon Ball Z
     
  </h2>
  <div
    className="card-grid"
  >
    <GifGridItem
      id="123"
      key="123"
      title="La Vieja PEYA"
      url="https://localhost/tumama/enpantaletas.jp"
    />
    <GifGridItem
      id="123"
      key="123"
      title="La Vieja PEYA"
      url="https://localhost/tumama/enpantaletas.jp"
    />
  </div>
</Fragment>
`;

exports[`Pruebas en componente <GifGrid /> Debe mostrar el componente <GifGrid /> correctamente 1`] = `
<Fragment>
  <h2
    className="animate__animated animate__fadeIn"
  >
     
    Dragon Ball Z
     
  </h2>
  <p
    className="animate__animated animate__flash"
  >
    Cargando...
  </p>
  <div
    className="card-grid"
  />
</Fragment>
`;
